File: Backjoon9663/Backjoon9663.c
#include <stdio.h>
#include <stdlib.h>

int count = 0;

int isPromising(int* q, int n) { 
    for (int i = 0; i < n; i++) {
        if (q[i] == q[n]) return 0;   // 같은 열인지
        if ((q[i] - q[n]) == (n - i)) return 0;   // '\' 방향
        if ((q[n] - q[i]) == (n - i)) return 0;   // '/' 방향
    }
    return 1;
}
void enumerate(int* q, int n, int N) { //q[] -> 해당 인덱스 행에서 놓은 위치, n -> 채워진 행 개수
    if (n == N)       // n이 끝까지 돌았다면 카운트를 증가한다.
        count++;
    else 
        for (int i = 0; i < N; i++) {
            q[n] = i; 
            if (isPromising(q, n) == 1) enumerate(q, n + 1, N);   // 유망하다면 계속 탐색(재귀호출)
        }
}

int main(){
    int N = 8;
    int*q = (int*)malloc(sizeof(int) * N);
    enumerate(q, 0, N);
    printf("%d", count);
}

File: Backjoon9663/Backjoon9663.java
package Backjoon9663;

public class Backjoon9663 {
    static int count = 0;
    public static boolean isPromising(int[] q, int n) {
        for (int i = 0; i < n; i++) {
            if (q[i] == q[n]) return false;   // 같은 열인지
            if ((q[i] - q[n]) == (n - i)) return false;   // '\' 방향
            if ((q[n] - q[i]) == (n - i)) return false;   // '/' 방향
        }
        return true;
    }
    public static void enumerate(int N) {
        int[] a = new int[N];
        enumerate(a, 0);
    }
    public static void enumerate(int[] q, int n) {   //q[] -> 해당 인덱스 행에서 놓은 위치, n -> 채워진 행 개수
        int N = q.length;
        if (n == N) {       // n이 끝까지 돌았다면 카운트를 증가한다.
            count++;
        } else {
            for (int i = 0; i < N; i++) {
                q[n] = i;
                if (isPromising(q, n)) enumerate(q, n + 1);   // 유망하다면 계속 탐색(재귀호출)
            }
        }
    }
    public static void main(String[] args) {
        int N = 8;
        enumerate(N);
        System.out.println(count);
    }
}